### ISSUE TITLE
Support mermaid diagrams

### ISSUE THREAD

**Issue by jyasskin** [speced/bikeshed#2539]
Either Mermaid or PlantUML would work, but Github picked Mermaid, so it probably makes sense to follow them.

**Comment by tabatkins** [comment]
Yeah, if there's a Python library for such I'd love to support this, akin to my support for railroad diagrams.

Right now, afaict, that isn't the case - I'd have to link in the mermaid JS lib, and it would result in document shift as the diagrams are rendered.

**Comment by tabatkins** [comment]
...well, hm, there is <https://mermaid.js.org/config/Tutorials.html#jupyter-integration-with-mermaid-js> which shows off using a live-rendering service to generate the images directly. I suppose that works, especially if I do some content-aware caching to avoid pinging it when you've already generated a given image.

**Comment by tabatkins** [comment]
I guess it would have to be a `bikeshed source` command, tho, which requires you to run things locally. `bikeshed spec` can't easily return a bunch of images alongside your spec. Tho I could embed it as base-64 instead, hmmmmmm.

**Comment by jyasskin** [comment]
Also https://github.com/mermaid-js/mermaid-cli#run-with-npx when Bikeshed notices that `npx` is available.

**Comment by tabatkins** [comment]
Ooooooh, I didn't see the `/svg` option, that's neat.

**Comment by sirex** [comment]
There is a command line interface to generate SVG files from Mermaid: https://github.com/mermaid-js/mermaid-cli

    mmdc -i input.mmd -o output.png -t dark -b transparent

I guess, Bikeshed could save Mermaid code snippets to a separate file and generate SVG files, that can be embedded into the output.

There is also an option to replace Mermaid code block with links to images directly in Markdown files:

    mmdc -i readme.template.md -o readme.md

**Comment by kainino0x** [comment]
WebGPU uses Mermaid to generate diagrams as part of our build. Mermaid does not have any themes that support `prefers-color-scheme`, unfortunately, so we just generate them in dark mode always.
Here's pointers on how we do it:
https://github.com/gpuweb/gpuweb/blob/059d5d920738befb70ac249af58d4eb2502d5ad7/spec/index.bs#L3264
https://github.com/gpuweb/gpuweb/blob/059d5d920738befb70ac249af58d4eb2502d5ad7/spec/Makefile#L19-L23
https://github.com/gpuweb/gpuweb/blob/059d5d920738befb70ac249af58d4eb2502d5ad7/tools/invoke-mermaid.sh#L11C1-L11C159
https://github.com/gpuweb/gpuweb/blob/059d5d920738befb70ac249af58d4eb2502d5ad7/tools/mermaid.json#L2-L5

I will note that I've found Mermaid's capabilities to be quite limiting compared to PlantUML. But it's generally serviceable.
